src/CommunityToolkit.Aspire.Hosting.Ollama/OllamaResourceBuilderExtensions.Model.cs

```diff
@@ -142,39 +141,136 @@ private static OllamaModelResource AddModelResourceCommand(
         string? confirmationMessage = null,
         string? iconName = null,
         IconVariant? iconVariant = IconVariant.Filled,
-        bool isHighlighted = false)
+        bool isHighlighted = false) =>
+            builder.WithCommand(
+                name: name,
+                displayName: displayName,
+                updateState: context =>
+                    context.ResourceSnapshot.State?.Text == KnownResourceStates.Running ?
+                        ResourceCommandState.Enabled :
+                        ResourceCommandState.Disabled,
+                executeCommand: async context =>
+                {
+                    var modelResource = builder.Resource;
+                    (var success, var endpoint) = await OllamaUtilities.TryGetEndpointAsync(modelResource, context.CancellationToken);
+
+                    if (!success || endpoint is null)
+                    {
+                        return new ExecuteCommandResult { Success = false, ErrorMessage = "Invalid connection string" };
+                    }
+
+                    var ollamaClient = new OllamaApiClient(endpoint);
+                    var logger = context.ServiceProvider.GetRequiredService<ResourceLoggerService>().GetLogger(modelResource);
+                    var notificationService = context.ServiceProvider.GetRequiredService<ResourceNotificationService>();
+
+                    return await executeCommand(modelResource, ollamaClient, logger, notificationService, context.CancellationToken);
+                },
+                displayDescription: displayDescription,
+                parameter: parameter,
+                confirmationMessage: confirmationMessage,
+                iconName: iconName,
+                iconVariant: iconVariant,
+                isHighlighted: isHighlighted
+            );
+
+    private static IResourceBuilder<OllamaModelResource> WithModelDownload(this IResourceBuilder<OllamaModelResource> builder)
     {
-        modelResource.Annotations.Add(new ResourceCommandAnnotation(
-            name: name,
-            displayName: displayName,
-            updateState: context =>
-                context.ResourceSnapshot.State?.Text == KnownResourceStates.Running ?
-                    ResourceCommandState.Enabled :
-                    ResourceCommandState.Disabled,
-            executeCommand: async context =>
+        builder.ApplicationBuilder.Eventing.Subscribe<ResourceReadyEvent>(builder.Resource.Parent, (@event, cancellationToken) =>
```

Comment:
Can we subscribe to the `ResourceReadyEvent` for a resource multiple times?
What happens when we have various models?

Reply:
Yes we can, they will fire in sequence though, which is why I use `Task.Run` to execute the download, spawning a background job to download the model so that the event handler becomes non-blocking (originally I didn't do that and it would be blocking, which is not ideal).